oeps/oep-0005.rst

```diff
@@ -0,0 +1,187 @@
+the infrastructure that was built out for installing Open edX in production.
+However, there are several problems with the current monolithic `Vagrant`_ image
+which we would like to address:
+
+1. As new services are added to the Open edX ecosystem, the `Vagrant`_ image
+   grows in size, requiring more system memory from the host machine.
+2. `Vagrant`_ imposes noticeable overhead during filesystem access, which especially
+   hurts any development that it doing file-heavy operations (such as manipulating
+   static assets).
+3. The monolithic `Vagrant`_ image isn't particularly representative of the
+   production service architecture, because all of the services are installed
+   on the same machine, which makes it harder for developers to be confident
+   that their changes will work when deployed.
+4. The shared services dependencies on things like MySQL and Elasticsearch mean
+   that all of the services in Devstack have to be upgraded at the same time,
+   or have to maintain compatibility with multiple database versions until all
+   services are ready to update. This hinders independent development on the
+   services.
+5. Because all of the services are installed on the same image, the image either
+   has to be updated by every user after downloading (to update to the latest
+   code versions), or the image has to be rebuilt every time any of the installed
+   services change (which results in lots of churn).
+6. Because the `Vagrant`_ image contains many separate services, the definition
+   of that image has to be separate from any of the service repositories, which
+   makes it hard to update (and to push updates out to users).
+
+
+Specification
+=============
+
+Interface
+~~~~~~~~~
+
+The primary developer interaction with ``devstack`` will be through ``make``
+shortcuts. The following targets will be standard::
+
+    make devstack.watch  # Start the devstack services (if they aren't already)
+                         # and display their logs on stdout
+
+    make devstack.shell  # Start the devstack services (if they aren't already)
+                         # and access a shell on the primary container
+                         # (as the user that owns the service, in the directory
+                         # that has the service source code)
+
+    make devstack.clean  # Stop the devstack services and remove any intermediate
+                         # data that they stored in the .devstack/ directory
+
+    make devstack.*      # Start the devstack services (if they aren't already)
+                         # and run `make *` from the code directory of the service
+                         # (as the service owner)
+
+Implementation
+~~~~~~~~~~~~~~
+
+Requirements
+------------
+
+1. Services that aren't part of the Open edX ecosystem must be run using
+   their own published containers (where available).
+2. Open edX containers must run all of their processes via `supervisor`_ to
```

Comment:
I was mostly just bringing up the types of questions that I don't see guidance for in the OEP. Happy to make whatever decisions I want for these various things, just wanted to bringing up some of the decision points I've had to face in the past to see if any were worth explicitly providing guidance for.

Also is it worth calling out that we expect the containers to be built for development purposes. If they happen to work in production as well, that's a bonus, but it's a secondary objective and if any mutually exclusive decisions need to be made, we should choose the development friendly option. That is, we shouldn't prevent commits of Dockerfiles that won't work well in production. If someone wants to modify them in the future or invest extra effort to make them also work in prod, more power to them as long as they don't sacrifice our developer-facing goals.